### ISSUE TITLE
computational cost of training per Epoch

### ISSUE THREAD

**Issue by deep-matter** [szc19990412/TransMIL#23]
i would like to ask how much the training step takes per Epoch, i used your built model and i modified the PPGE model by adding FFT to reduce the dimension Convolution operation, only issue i noticed was that the Trainer took a lot of time to finish single Epoch, that's is related to size the shape of the image (2154,1024) or i missed something

**Comment by szc19990412** [comment]
> i would like to ask how much the training step takes per Epoch, i used your built model and i modified the PPGE model by adding FFT to reduce the dimension Convolution operation, only issue i noticed was that the Trainer took a lot of time to finish single Epoch, that's is related to size the shape of the image (2154,1024) or i missed something

Because we set the batch size to one, the training step in one Epoch is equal to the number of your training slide. Meanwhile, because we preprocess all the WSIs into features, we test the training is very quick in RTX3090, roughly 0.5 min per epoch if we have 400 slides.

**Comment by szc19990412** [comment]
> I trained on my own dataset(200train +30val)，each slide were cut into 500~1500 tiles, and then embeded into 2048 dim vectors. It cost about 1min per epoch. I wonder how many slide is in your dataset(train and val) and how long it takes per epoch. By the way, my result were pretty poor and my training procedure were not stable, don't know where maybe wrong.

1、Each slide has 500~1500 tiles, have you process the WSI in the 20x magnification or higher?  2、You can also test the performance of other MIL methods in your dataset, such as ABMIL or CLAM, if the task is challenging or the dataset is limited.

**Comment by hans0809** [comment]
> 

For each slide, I did 4 times downsample and then cut into 224x224 small patches.
I tried DTFD-MIL(《DTFD-MIL: Double-Tier Feature Distillation Multiple Instance Learning for
Histopathology Whole Slide Image Classification》), and got a better performance.

I print the softmax(pred_logits) for both TransMIL and DTFD-MIL, the latter is more discriminative:
`TransMIL`
[![zoykUP.png](https://s1.ax1x.com/2022/12/15/zoykUP.png)](https://imgse.com/i/zoykUP)
`DTFD-MIL`
[![zoyegg.png](https://s1.ax1x.com/2022/12/15/zoyegg.png)](https://imgse.com/i/zoyegg)

Maybe something wrong with my implementation...

**Comment by szc19990412** [comment]
This result seems strange, as it appears to show that the model is overfitting. Because the DTFD model is built on a smaller model ABMIL, you might experiment with Transformer aggregation in lower dimensional aggregation, such as lowering from 2048 to 128 or 256. Furthermore, have you used our Pytorch lightning framework and the Ranger optimizer.

**Comment by deep-matter** [comment]
> 

`python 
class FF_ATPEG(nn.Module):
    def __init__(self, modes1, modes2,  width):
        super(FF_ATPEG, self).__init__()
        """
        The overall network. It contains 4 layers of the Fourier layer.
        1. Lift the input to the desire channel dimension by self.fc0 .
        2. 4 layers of the integral operators u' = (W + K)(u).
            W defined by self.w; K defined by self.conv .
        3. Project from the channel space to the output space by self.fc1 and self.fc2 .
        
        input: the solution of the coefficient function and locations (a(x, y), x, y)
        input shape: (batchsize, x=s, y=s, c=3)
        output: the solution 
        output shape: (batchsize, x=s, y=s, c=1)
        """

        self.modes1 = modes1
        self.modes2 = modes2
        self.width = width
        self.attetion_gate = Attention_block()
        self.padding = 9 # pad the domain if input is non-periodic
        #self.fc0 = nn.Linear(512,512 ) # input channel is 3: (a(x, y), x, y)
        self.conv0  = SpectralConv2d(self.width, self.width, self.modes1, self.modes2)
        self.conv1 = SpectralConv2d(self.width, self.width, self.modes1, self.modes2)
        self.conv2 = SpectralConv2d(self.width, self.width, self.modes1, self.modes2)
        self.projection_= nn.Conv2d(self.width, self.width, 7, 1, 7//2, groups=self.width)
        self.projection_1 = nn.Conv2d(self.width, self.width, 5, 1, 5//2, groups=self.width)
        self.projection_2 = nn.Conv2d(self.width, self.width, 3, 1, 3//2, groups=self.width)
        
    def forward(self,x ,H, W):
            B,_,C = x.shape
            cls_token,feat_token = x[:,0],x[:, 1:]
            cnn_feat = feat_token.transpose(1,2).view(B,C,H,W)

            x1 =  self.conv0(cnn_feat) 
            x2 =  self.projection_(x1)
            x =  x1+x2
            x = F.gelu(x)

            x3 =  self.conv1(cnn_feat) 
            x4 =  self.projection_1(x3)
            x =self.attetion_gate(x4,x1)## x1+x2
            x = F.gelu(x)

            x5 =  self.conv2(x) 
            x6 =  self.projection_2(x5)
            x = self.attetion_gate(x6,x3) # ## x1+x2
            x = F.gelu(x)

            x = x.flatten(2).transpose(1,2)
            x = torch.cat((cls_token.unsqueeze(1),x),dim=1)       
            return x`

**Comment by JD910** [comment]
> 
Nice work.

Any recommendations to preprocess the WSIs into features (since the quality of features may influence the classification directly)?
